lib/src/voip/group_call_session.dart

```diff
@@ -0,0 +1,1361 @@
+/*
+ *   Famedly Matrix SDK
+ *   Copyright (C) 2021 Famedly GmbH
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU Affero General License as
+ *   published by the Free Software Foundation, either version 3 of the
+ *   License, or (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *   GNU Affero General License for more details.
+ *
+ *   You should have received a copy of the GNU Affero General License
+ *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:core';
+import 'dart:typed_data';
+
+import 'package:collection/collection.dart';
+import 'package:webrtc_interface/webrtc_interface.dart';
+
+import 'package:matrix/matrix.dart';
+import 'package:matrix/src/utils/cached_stream_controller.dart';
+import 'package:matrix/src/utils/crypto/crypto.dart';
+import 'package:matrix/src/voip/models/call_membership.dart';
+import 'package:matrix/src/voip/models/call_options.dart';
+import 'package:matrix/src/voip/models/voip_id.dart';
+import 'package:matrix/src/voip/utils/stream_helper.dart';
+
+/// Holds methods for managing a group call. This class is also responsible for
+/// holding and managing the individual `CallSession`s in a group call.
+class GroupCallSession {
+  // Config
+  static const updateExpireTsTimerDuration = Duration(seconds: 15);
+  static const expireTsBumpDuration = Duration(seconds: 45);
+  static const activeSpeakerInterval = Duration(seconds: 5);
+
+  final Client client;
+  final VoIP voip;
+  final Room room;
+
+  /// is a list of backend to allow passing multiple backends in the future
+  /// we use the first backend everywhere as of now
+  final List<CallBackend> backends;
+  final String? application;
+  final String? scope;
+
+  GroupCallState state = GroupCallState.localCallFeedUninitialized;
+  StreamSubscription<CallSession>? _callSubscription;
+
+  /// participant:volume
+  final Map<CallParticipant, double> audioLevelsMap = {};
+  CallParticipant? activeSpeaker;
+  WrappedMediaStream? localUserMediaStream;
+  WrappedMediaStream? localScreenshareStream;
+  String? localDesktopCapturerSourceId;
+  List<CallSession> callSessions = [];
+
+  CallParticipant? get localParticipant => voip.localParticipant;
+
+  /// userId:deviceId
+  List<CallParticipant> participants = [];
+  List<WrappedMediaStream> userMediaStreams = [];
+  List<WrappedMediaStream> screenshareStreams = [];
+  late String groupCallId;
```

Comment:
I don't think this needs to be late, but you might need a factory constructor then?